Updates for new interfaces

NerdDiffer opened the pull request:
Hi, I was interested in contributing to this API client. 

It's been a few months since Backpack.tf added some interfaces and upgraded a version of one of them, so I made some updates. I also made a few changes to the way query string parameters are set. Please review and let me know what you think.

Thanks,
Rafael

rafaelespinoza committed:
Add two new API interfaces, helper module & tests

In the last few months, the backpack.tf site started offering two new
interfaces ('IGetMarketPrices' & 'IGetPriceHistory'). This commit
addresses those.

Since the 'IGetPriceHistory' interface requires some more parameters in
the request, I added a dependency on lodash's assign module. This
merges user input with the 'key' and 'appid' parameters before it is
sent to the 'querystring.stringify' method.

Add a module to help set query string parameters.

Add some tests and fixtures.

rafaelespinoza committed:
Move a helper method inside of constructor function

Set some boilerplate values (the client's api key and the appid) from
inside of the constructor function instead of using the helper.

Make use of the 'assign' module when there is some input needed, like
in the getUserListings() and getPrices() methods.

Note: I changed order of parameters in the test for 'getPrices()'
function on line 36, in order for it to pass.

rafaelespinoza committed:
Update README and update interface version for IGetUserListings

Update README with some tips on how to use the new functions. Update
the version on the endpoint for IGetUserListings.

KenanY commented on index.js:
```diff
@@ -15,20 +18,20 @@ function backpacktf(apiKey, app) {
-  this.app = app || 440;
+  this.basicOptions = {
+    key: apiKey,
+    appid: app || 440,
```
Trailing comma.

KenanY commented:
Am I wrong in saying that the use of `lodash.assign` would mutate `this.basicOptions` and potentially affect future function calls?

KenanY commented:
Also I prefer single quotes to double quotes. I also think we can do with moving `getPriceHistoryOptions` to `index.js` instead of putting it in a new file.

This contribution is great though! Looking forward to merging :)

KenanY commented on test/helpers.js:
```diff
@@ -0,0 +1,31 @@
+    quality: 'Haunted',
+    tradable: '1',
+    craftable: '1',
+    priceindex: '0',
```
Trailing comma.

rafaelespinoza committed:
Don't mutate basicOptions. Squashed commit of the following:

commit e6962b388ca343b48b48c0c9b0b99af15ca151bb
Author: Rafael Espinoza <rafael@rafaelespinoza.com>
Date:   Mon Aug 31 14:05:02 2015 -0700

    Add another helper method to set basicOptions

    Add dependency on 'lodash.clone'.

    Using 'lodash.assign' by itself was mutating the original values of
    'this.basicOptions'.

    So, a new wrapper method clones it and then assigns  any new
    key/value pairs to the clone. Hopefully, this keeps the query
    parameters slim throughout successive API calls.

    It seems important to test this, but it doesn't need public exposure.
    So I thought keeping the 'helpers' module might be a good idea.

commit 507e8e144eaa3fce4db9a4ea8871ae0471d2827e
Author: Rafael Espinoza <rafael@rafaelespinoza.com>
Date:   Mon Aug 31 11:21:13 2015 -0700

    Remove trailing commas and change double-quotes to single-quotes

NerdDiffer commented:
Thanks for the feedback. You are right, calling `lodash.assign` on the `basicOptions` property will mutate it. I wrote some tests to confirm this.

So, I'm a little confused. Did you need me to make some of these changes? Just in case, I have something pushed to my fork.

It seems kind of important to test the setting of query string parameters (or...at least, what goes into it), but without mutating the original value of `basicOptions`. I don't think it needs public exposure though. So, I thought keeping the helper module might be a good idea, if only to make testing easier.

NerdDiffer commented:
I should add, I'm not sure how to design it so these, effectively private, methods are testable. I'm pretty sure that they're important enough to be tested, but I may be wrong. But also, the two methods in the `helper` module don't need to be directly called by the client. Therefore, they are not exported as part of the backpacktf object. 

Is keeping those 'private-ish' methods in a separate module the best way to do this?

KenanY commented:
@NerdDiffer Amended, squashed, and landed as https://github.com/KenanY/backpack.tf/commit/f2b0e22d757c27677b7e70556679b02436285ff9. Thank you so much for this contribution! I changed many little-things during my amending process but you still retain all authorship. In a better world, I should have merged this as several commits:

  - update `IGetUserListings` endpoint
  - add `IGetMarketPrices`
  - add `IGetPriceHistory`
  - fix `getPrices()` test
  - etc.

but once I had the monolithic commit I relented to just merging that :)

I have also added you as a collaborator of this repository. But let's please continue using pull requests for the sake of reviewing each other's code :)

NerdDiffer commented:
@KenanY, Sounds good. Thanks! I'll use pull requests in the future, and I like your use of the `assign` module to handle the `basicOptions` property.

For future reference, would it have been easier if I had spread out the proposed changes among more commits?

KenanY commented:
@NerdDiffer Yes, more commits/pull requests would result in smaller diffs to review :)
